integrations/mcp/src/haystack_integrations/tools/mcp/mcp_toolset.py

```diff
@@ -0,0 +1,232 @@
+    This dual connectivity allows for integrating with both remote and local MCP servers.
+
+    Example using MCPToolset in a Haystack Pipeline:
+    ```python
+    import os
+    from haystack import Pipeline
+    from haystack.components.converters import OutputAdapter
+    from haystack.components.generators.chat import OpenAIChatGenerator
+    from haystack.components.tools import ToolInvoker
+    from haystack.dataclasses import ChatMessage
+    from haystack_integrations.tools.mcp import MCPToolset, StdioServerInfo
+
+    # Create server info for the time service (can also use SSEServerInfo for remote servers)
+    server_info = StdioServerInfo(command="uvx", args=["mcp-server-time", "--local-timezone=Europe/Berlin"])
+
+    # Create the toolset - this will automatically discover all available tools
+    # You can optionally specify which tools to include
+    mcp_toolset = MCPToolset(
+        server_info=server_info,
+        tool_names=["get_current_time"]  # Only include the get_current_time tool
+    )
+
+    # Create a pipeline with the toolset
+    pipeline = Pipeline()
+    pipeline.add_component("llm", OpenAIChatGenerator(model="gpt-4o-mini", tools=mcp_toolset))
+    pipeline.add_component("tool_invoker", ToolInvoker(tools=mcp_toolset))
+    pipeline.add_component(
+        "adapter",
+        OutputAdapter(
+            template="{{ initial_msg + initial_tool_messages + tool_messages }}",
+            output_type=list[ChatMessage],
+            unsafe=True,
+        ),
+    )
+    pipeline.add_component("response_llm", OpenAIChatGenerator(model="gpt-4o-mini"))
+    pipeline.connect("llm.replies", "tool_invoker.messages")
+    pipeline.connect("llm.replies", "adapter.initial_tool_messages")
+    pipeline.connect("tool_invoker.tool_messages", "adapter.tool_messages")
+    pipeline.connect("adapter.output", "response_llm.messages")
+
+    # Run the pipeline with a user question
+    user_input = "What is the time in New York? Be brief."
+    user_input_msg = ChatMessage.from_user(text=user_input)
+
+    result = pipeline.run({"llm": {"messages": [user_input_msg]}, "adapter": {"initial_msg": [user_input_msg]}})
+    print(result["response_llm"]["replies"][0].text)
+    ```
+
+    You can also use the toolset via MCP SSE to talk to remote servers:
+    ```python
+    from haystack_integrations.tools.mcp import MCPToolset, SSEServerInfo
+    from haystack.components.tools import ToolInvoker
+
+    # Create the toolset with an SSE connection
+    sse_toolset = MCPToolset(
+        server_info=SSEServerInfo(base_url="http://some-remote-server.com:8000"),
+        tool_names=["add", "subtract"]  # Only include specific tools
+    )
+
+    # Use the toolset as shown in the pipeline example above
+    ```
+    """
+
+    def __init__(
+        self,
+        server_info: MCPServerInfo,
+        tool_names: list[str] | None = None,
+        connection_timeout: int = 30,
+        invocation_timeout: int = 30,
```

Comment:
The timeouts are of type float not int. E.g. Your run method of `AsyncExecutor` has this signature
```python
def run(self, coro: Coroutine[Any, Any, Any], timeout: float | None = None) -> Any:
```